src/RustyObjectStore.jl

```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
+- `storage_account_key::Option{String}`: (Optional) Azure storage account key.
+- `storage_sas_token::Option{String}`: (Optional) Azure storage SAS token.
+- `host::Option{String}`: (Optional) Alternative Azure host. For example, if using Azurite.
+- `opts::ClientOptions`: (Optional) Client configuration options.
+"""
+struct AzureConfig <: AsConfig
+    config::Config
+    function AzureConfig(;
+        storage_account_name::String,
+        container_name::String,
+        storage_account_key::Option{String} = nothing,
+        storage_sas_token::Option{String} = nothing,
+        host::Option{String} = nothing,
+        opts::ClientOptions = ClientOptions()
+    )
+        (
+            !isnothing(storage_account_key)
+            && !isnothing(storage_sas_token)
+        ) && error("Should provide either a storage_account_key or a storage_sas_token")
+
+        params = copy(opts.params)
+
+        params["azure_storage_account_name"] = storage_account_name
+        params["azure_container_name"] = container_name
+
+        if !isnothing(storage_account_key)
+            params["azure_storage_account_key"] = storage_account_key
+        end
+
+        if !isnothing(storage_sas_token)
+            params["azure_storage_sas_token"] = storage_sas_token
+        end
+
+        if !isnothing(host)
+            params["azurite_host"] = host
+        end
+
+        return new(Config("az://$(container_name)/", params))
+    end
+end
+
+as_config(wrapper::AzureConfig) = wrapper.config
+
+"""
+    $TYPEDEF
+
+# Keyword Arguments
+- `region::String`: AWS S3 region.
+- `bucket_name::String`: AWS S3 bucket name.
+- `access_key_id::Option{String}`: (Optional) AWS S3 access key id.
+- `secret_access_key::Option{String}`: (Optional) AWS S3 secret access key.
+- `session_token::Option{String}`: (Optional) AWS S3 session_token.
+- `host::Option{String}`: (Optional) Alternative S3 host. For example, if using Minio.
+- `opts::ClientOptions`: (Optional) Client configuration options.
+"""
+struct AwsConfig <: AsConfig
+    config::Config
+    function AwsConfig(;
+        region::String,
+        bucket_name::String,
+        access_key_id::Option{String} = nothing,
+        secret_access_key::Option{String} = nothing,
+        session_token::Option{String} = nothing,
+        host::Option{String} = nothing,
+        opts::ClientOptions = ClientOptions()
+    )
+        params = copy(opts.params)
+
+        params["region"] = region
+        params["bucket_name"] = bucket_name
+
+        if !isnothing(access_key_id)
+            params["aws_access_key_id"] = access_key_id
+        end
+
+        if !isnothing(secret_access_key)
+            params["aws_secret_access_key"] = secret_access_key
+        end
+
+        if !isnothing(session_token)
+            params["aws_session_token"] = session_token
+        end
+
+        if !isnothing(host)
+            params["minio_host"] = host
+        end
+
+        return new(Config("s3://$(bucket_name)/", params))
```

Comment:
why do we "eagerly" turn all the given params into a single `String`, rather than storing the data as given and then just turning it into a string when needed for a `get!` or `put` call

I presume for efficiency (i.e. so we only need to create each config string once, no matter how many `get!`/`put` calls)

Can we add that to a comment somewhere? If that's an important consideration, i think we should document it so it never gets refactored away

Reply:
Yes, it is still debatable if this will end up being significant in terms of performance (mainly in the future). But currently we are doing this for efficiency as you correctly presumed.

With the string stored on the struct we just need to send its pointer to Rust and hash its contents there to find the correct client for the request.

Storing the data as given would be much nicer for sure, but we would need to pay for serialization and allocation for each request (for 2MB requests it certainly does not matter much, but who knows if we may need to do smaller ones in the future).

Reply:
yeah, makes sense -- just worth adding a comment about this somewhere i think